### DISCUSSION TITLE
System load widget

### DISCUSSION THREAD

**Discussion by DCsunset** [qtile/qtile#3561] [Ideas]
It would be great to add a widget to show the system load average, similar to the CPU and memory widgets.

**Comment by elParaguayo** [comment]
Which command are you using to show load? Is this `uptime`?

**Reply by DCsunset** [reply]
Yes, it is. I think  `psutil` library can also be used to get the stat.

**Reply by m-col** [reply]
What is system load, if not CPU load?

**Reply by DCsunset** [reply]
It is CPU load and it reflects the CPU queue length instead of CPU utilization, which can directly reflect the amount of load on a CPU.

Now the CPU widget only shows the CPU utilization, so it would be good to see the CPU load as well.

**Reply by m-col** [reply]
Ah cool. Would it make sense to have it as part of the CPU widgets then?

**Reply by DCsunset** [reply]
From my experience, most bars usually make it a separate widget because the stat is retrieved from another command. But I think it's okay to have it as part of the CPU widget as well.

**Comment by SweepingBishops** [comment]
Should be an easy one. I'll work on it, unless you've already started elParaguayo.

**Reply by elParaguayo** [reply]
That would be great. I hadn't done anything other than look at the output from `psutil`.  

I assume you'd just use the 1 min load average number.

**Reply by SweepingBishops** [reply]
I was thinking of using `uptime` and a bit of string slicing.
Also show the 1 min one by default and cycle through them with mouse clicks.

**Reply by elParaguayo** [reply]
The thing that bothers me about the graphs is that there's no scale. It's just adjusted to the current maximum value. This means you've got no idea of what the graph really represents, other than relative comparisons over the sample period.

I might look at adding the ability to set maximum values, having different colours when values go over certain limits etc. 

That's separate to this work though!

**Reply by SweepingBishops** [reply]
Yeah, and that will probably be a _bit_ advanced for me. :smile:

**Reply by elParaguayo** [reply]
> I was thinking of using `uptime` and a bit of string slicing. Also show the 1 min one by default and cycle through them with mouse clicks.

I'd use `psutil.getloadavg()` ([docs](https://psutil.readthedocs.io/en/latest/#psutil.getloadavg)) - then you can avoid string splitting.

The `graph.py` file already imports `psutil` as well so it's not a new dependency for graphs.

**Reply by SweepingBishops** [reply]
Well I wasn't thinking of graphs. I was talking about a text widget. Does it have to be a graph?

**Reply by elParaguayo** [reply]
Ah! Ok. Good point.

You're absolutely right - I'd read the first post as being about graphs because both of the widgets mentioned also have graphs but that was never mentioned. A text widget probably makes more sense here.
